### PR TITLE
HAST-327: Update Lombiq.Analyzers to latest

### LINKED ISSUE #117: Update Lombiq.Analyzers to latest (HAST-327)
...and fix and new code analyzer violations. Note the breaking changes: https://github.com/Lombiq/.NET-Analyzers/releases/tag/v4.0.0

[Jira issue](https://lombiq.atlassian.net/browse/HAST-327)

### PR DESCRIPTION
**Pull request by sarahelsaig**
[HAST-327](https://lombiq.atlassian.net/browse/HAST-327)
Fixes #117

[HAST-327]: https://lombiq.atlassian.net/browse/HAST-327?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: .globalconfig [hunk 1]
```diff
@@ -0,0 +1,10 @@
+is_global = true
+global_level = 10
+
+# Code analysis rules for Hastlayer, extending Lombiq.Analyzers.globalconfig.
+
+# Microsoft.CodeAnalysis.NetAnalyzers rules
+dotnet_diagnostic.CA1062.severity = silent
+dotnet_diagnostic.CA1724.severity = silent
+dotnet_diagnostic.CA1848.severity = none
+dotnet_diagnostic.CA2007.severity = silent
```
Comment (Piedone) [inline .globalconfig]: Document why we disable these. Also, try `none` for all of these (unless you indeed want them to be silent only); previously, some rules got disabled only with `silent` for some weird reason, but that might not be the case now.

File: .globalconfig [hunk 2]
```diff
@@ -0,0 +1,10 @@
+is_global = true
+global_level = 10
```
Comment (Piedone) [inline .globalconfig]: This shouldn't be necessary.

File: test/TestInputAssemblies/Hast.TestInputs.Static/ObjectUsingCases.cs [hunk 3]
```diff
@@ -1,13 +1,15 @@
+using System.Diagnostics.CodeAnalysis;
+
 namespace Hast.TestInputs.Static;
 
 public class ObjectUsingCases
 {
+    [SuppressMessage("Style", "IDE0270:Use coalesce expression", Justification = "We want to test this specific syntax.")]
+    [SuppressMessage("Style", "IDE0074:Use compound assignment", Justification = "Same.")]
+    [SuppressMessage("Minor Code Smell", "S3240:The simplest possible condition syntax should be used", Justification = "Same.")]
```
Comment (Piedone) [inline test/TestInputAssemblies/Hast.TestInputs.Static/ObjectUsingCases.cs:9]: The two `#pragma warning restore`s are still there.

File: src/Hastlayer/Hast.Algorithms/Fix64.cs [hunk 4]
```diff
@@ -104,13 +104,7 @@ public int[] ToIntegers()
...
     /// it is negative.
     /// </summary>
-    public static int Sign(Fix64 value) =>
-        value.RawValue switch
-        {
-            { } when value.RawValue < 0 => -1,
-            { } when value.RawValue > 0 => 1,
-            _ => 0,
-        };
+    public static int Sign(Fix64 value) => Math.Sign(value.RawValue);
```
Comment (Piedone) [inline src/Hastlayer/Hast.Algorithms/Fix64.cs]: This also needs to be transformable, what the `Math` method not necessarily is.

File: src/Hastlayer/Hast.Catapult/CatapultLibrary.cs [hunk 5]
```diff
@@ -551,4 +547,9 @@ private Task TesterWriteLineAsync(string format, params object[] objects)
         LogFunction?.Invoke((uint)level, text.ToString(CultureInfo.InvariantCulture) + Environment.NewLine);
 
     public override string ToString() => InstanceName;
+
+    private static object GetOrDefault(IDictionary<string, object> config, string key, object fallback = null) =>
+        config.TryGetValue(key, out var result)
+            ? result ?? fallback
+            : fallback;
```
Comment (Piedone) [inline src/Hastlayer/Hast.Catapult/CatapultLibrary.cs]: This looks like a good candidate for HL.

File: src/Hastlayer/Hast.Communication/Services/FpgaIpEndpointFinder.cs [hunk 6]
```diff
@@ -45,21 +49,21 @@ public async Task<IEnumerable<IFpgaEndpoint>> FindFpgaEndpointsAsync()
             var endpoints = new List<IPEndPoint>();
 
             // Send request to all broadcast addresses on all the supported network interfaces.
-            foreach (var suppertedNetworkInterface in NetworkInterface.GetAllNetworkInterfaces()
```
Comment (Piedone) [inline src/Hastlayer/Hast.Communication/Services/FpgaIpEndpointFinder.cs:48]: image

File: src/Hastlayer/Hast.Communication/Services/FpgaIpEndpointFinder.cs [hunk 7]
```diff
@@ -28,10 +29,13 @@ public FpgaIpEndpointFinder(IClock clock, ILogger<FpgaIpEndpointFinder> logger)
         _logger = logger;
     }
 
+    [SuppressMessage(
+        "Style",
+        "IDE0230:Use UTF-8 string literal",
+        Justification = "False positive, shouldn't replace the char constant with a literal.")]
```
Comment (Piedone) [inline src/Hastlayer/Hast.Communication/Services/FpgaIpEndpointFinder.cs:36]: This is flagged as unnecessary for me.

File: src/Hastlayer/Hast.Transformer/Services/ConstantValuesSubstitution/ConstantValuesSubstitutingVisitor.cs [hunk 8]
```diff
@@ -421,7 +421,7 @@ public override void VisitAssignmentExpression(AssignmentExpression assignmentEx
             {
                 ConstantValuesTable.MarkAsNonConstant(assignmentExpression.Left, _constructor);
             }
-            else if (right != null)
```
Comment (Piedone) [inline src/Hastlayer/Hast.Transformer/Services/ConstantValuesSubstitution/ConstantValuesSubstitutingVisitor.cs:424]: Why can this be removed?

File: src/Hastlayer/Hast.Vitis/Services/VitisHardwareImplementationComposerBuildProvider.cs [hunk 9]
```diff
@@ -251,16 +252,9 @@ public Task CleanupAsync(IHardwareImplementationCompositionContext context)
...
-        {
-            throw new FileNotFoundException(
-                "Unable to find the platform xpfm file. The supported platforms are: " +
-                string.Join(", ", deviceManifest.SupportedPlatforms));
-        }
-
-        return device;
+            .FullName ?? throw new FileNotFoundException(
+            "Unable to find the platform xpfm file. The supported platforms are: " +
+            string.Join(", ", deviceManifest.SupportedPlatforms));
```
Comment (Piedone) [inline src/Hastlayer/Hast.Vitis/Services/VitisHardwareImplementationComposerBuildProvider.cs]:
```suggestion
            .FullName ??
            throw new FileNotFoundException(
                "Unable to find the platform xpfm file. The supported platforms are: " +
                string.Join(", ", deviceManifest.SupportedPlatforms));
```

**Review by Piedone — CHANGES_REQUESTED** [review]

File: .globalconfig (hunk 1 above)
Comment (sarahelsaig) [reply .globalconfig]: why aren't they documented in Lombiq.Analyzers.OrchardCore though?

Comment (Piedone) [reply .globalconfig]: Apparently, that reviewer wasn't attentive enough :D.

File: src/Hastlayer/Hast.Communication/Services/FpgaIpEndpointFinder.cs (hunk 7 above)
Comment (sarahelsaig) [reply src/Hastlayer/Hast.Communication/Services/FpgaIpEndpointFinder.cs:36]: VS being garbage again.

File: src/Hastlayer/Hast.Catapult/CatapultLibrary.cs (hunk 5 above)
Comment (sarahelsaig) [reply src/Hastlayer/Hast.Catapult/CatapultLibrary.cs]: On second thought 2 of the 3 usages is just `config.GetMaybe(key)` and the other is `config.GetMaybe(key) ?? fallback` so no need for a separate method.

File: src/Hastlayer/Hast.Transformer/Services/ConstantValuesSubstitution/ConstantValuesSubstitutingVisitor.cs (hunk 8 above)
Comment (sarahelsaig) [reply src/Hastlayer/Hast.Transformer/Services/ConstantValuesSubstitution/ConstantValuesSubstitutingVisitor.cs:424]: If `right` was null then the `assignmentExpression.Right is not PrimitiveExpression right` in the branch above would've triggered. So this condition is always true.

File: src/Hastlayer/Hast.Communication/Services/FpgaIpEndpointFinder.cs (hunk 7 above)
Comment (Piedone) [reply src/Hastlayer/Hast.Communication/Services/FpgaIpEndpointFinder.cs:36]: image

Comment (sarahelsaig) [reply src/Hastlayer/Hast.Communication/Services/FpgaIpEndpointFinder.cs:36]: My what?
image

File: .globalconfig (hunk 1 above)
Comment (sarahelsaig) [reply .globalconfig]: And what would you have written for disabling `CA2007 : Consider calling ConfigureAwait on the awaited task`? I've documented the rest, but I have a hard time articulating this one.

Comment (Piedone) [reply .globalconfig]: OC doesn't have a GUI like a desktop app, that's commonly used to prevent thread deadlocks there.

**Review by Piedone — APPROVED** [review]

**Pull request merged**